crates/wast/src/binary.rs

```diff
@@ -1108,7 +1228,159 @@ fn find_names<'a>(
     return ret;
 }
 
-impl Names<'_> {
+#[derive(Default)]
+struct ComponentNames<'a> {
+    component: Option<&'a str>,
+    funcs: Vec<(u32, &'a str)>,
+    func_idx: u32,
+    locals: Vec<(u32, Vec<(u32, &'a str)>)>,
+    labels: Vec<(u32, Vec<(u32, &'a str)>)>,
+    components: Vec<(u32, &'a str)>,
+    component_idx: u32,
+    instances: Vec<(u32, &'a str)>,
+    instance_idx: u32,
+    types: Vec<(u32, &'a str)>,
+    type_idx: u32,
+}
+
+fn find_component_names<'a>(
+    component_id: &Option<Id<'a>>,
+    component_name: &Option<NameAnnotation<'a>>,
+    fields: &[ComponentField<'a>],
+) -> ComponentNames<'a> {
+    fn get_name<'a>(id: &Option<Id<'a>>, name: &Option<NameAnnotation<'a>>) -> Option<&'a str> {
+        name.as_ref().map(|n| n.name).or(id.and_then(|id| {
+            if id.is_gensym() {
+                None
+            } else {
+                Some(id.name())
+            }
+        }))
+    }
+
+    enum Name {
+        Type,
+        Func,
+        Module,
+        Component,
+        Instance,
+    }
+
+    let mut ret = ComponentNames::default();
+    ret.component = get_name(component_id, component_name);
+    for field in fields {
+        // Extract the kind/id/name from whatever kind of field this is...
+        let (kind, id, name) = match field {
+            ComponentField::Import(i) => (
+                match i.item.kind {
+                    ItemKind::Func(_) => Name::Func,
+                    ItemKind::Table(_) => panic!("table inside component"),
+                    ItemKind::Memory(_) => panic!("memory inside component"),
+                    ItemKind::Global(_) => panic!("global inside component"),
+                    ItemKind::Tag(_) => panic!("tag inside component"),
+                    ItemKind::Instance(_) => Name::Instance,
+                    ItemKind::Module(_) => Name::Module,
+                },
+                &i.item.id,
+                &i.item.name,
+            ),
+            ComponentField::NestedModule(m) => (Name::Component, &m.id, &m.name),
+            ComponentField::NestedComponent(c) => (Name::Component, &c.id, &c.name),
+            ComponentField::Instance(i) => (Name::Instance, &i.id, &i.name),
+            ComponentField::Type(t) => (Name::Type, &t.id, &t.name),
+            ComponentField::Func(f) => (Name::Func, &f.id, &f.name),
+            ComponentField::Alias(a) => (
+                match a.kind {
+                    ExportKind::Func => Name::Func,
+                    ExportKind::Table => panic!("table inside component"),
+                    ExportKind::Memory => panic!("memory inside component"),
+                    ExportKind::Global => panic!("global inside component"),
```

Comment:
I think these are valid cases for components, right? In that I think this should be handled rather than panicking?